src/erasure.rs

```diff
@@ -161,263 +168,418 @@ pub fn decode_blocks(data: &mut [&mut [u8]], coding: &[&[u8]], erasures: &[i32])
+
+        let coding_start = block_end - NUM_CODING;
+
+        for i in coding_start..block_end {
+            let n = i % window.len();
+            assert!(window[n].coding.is_none());
+
+            window[n].coding = Some(recycler.allocate());
+
+            let coding = window[n].coding.clone().unwrap();
+            let mut coding_wl = coding.write().unwrap();
+            for i in 0..max_data_size {
+                coding_wl.data[i] = 0;
             }
-            for b in &coding_blobs {
-                coding_locks.push(
-                    b.write()
-                        .expect("'coding_locks' arr in pub fn generate_coding"),
+            // copy index and id from the data blob
+            if let Some(data) = &window[n].data {
+                let data_rl = data.read().unwrap();
+
+                let index = data_rl.get_index().unwrap();
+                let id = data_rl.get_id().unwrap();
+
+                trace!(
+                    "{:x} copying index {} id {:?} from data to coding",
+                    debug_id,
+                    index,
+                    id
                 );
+                coding_wl.set_index(index).unwrap();
+                coding_wl.set_id(id).unwrap();
             }
-            for (i, l) in coding_locks.iter_mut().enumerate() {
-                trace!("i: {} coding: {} size: {}", i, l.data[0], max_data_size);
-                coding_ptrs.push(&mut l.data_mut()[..max_data_size]);
+            coding_wl.set_size(max_data_size);
+            if coding_wl.set_coding().is_err() {
+                return Err(ErasureError::EncodeError);
             }
 
-            generate_coding_blocks(coding_ptrs.as_mut_slice(), &data_ptrs)?;
-            debug!(
-                "consumed: {} data: {}:{} coding: {}:{}",
-                consumed,
-                block_start,
-                block_start + NUM_DATA,
-                coding_start,
-                coding_end
+            coding_blobs.push(coding.clone());
+        }
+
+        let mut data_locks = Vec::with_capacity(NUM_DATA);
+        for b in &data_blobs {
+            data_locks.push(
+                b.write()
+                    .expect("'data_locks' write lock in pub fn generate_coding"),
+            );
+        }
+
+        let mut data_ptrs: Vec<&[u8]> = Vec::with_capacity(NUM_DATA);
+        for (i, l) in data_locks.iter_mut().enumerate() {
+            trace!("{:x} i: {} data: {}", debug_id, i, l.data[0]);
+            data_ptrs.push(&l.data[..max_data_size]);
+        }
+
+        let mut coding_locks = Vec::with_capacity(NUM_CODING);
+        for b in &coding_blobs {
+            coding_locks.push(
+                b.write()
+                    .expect("'coding_locks' arr in pub fn generate_coding"),
             );
-            block_start += NUM_CODED;
         }
+
+        let mut coding_ptrs: Vec<&mut [u8]> = Vec::with_capacity(NUM_CODING);
+        for (i, l) in coding_locks.iter_mut().enumerate() {
```

Comment:
Can all 4 loops be changed to use `map`?

Reply:
if I use map() can I pre-allocate the destination vectors?

Reply:
Yes, you can expect collect() to do that for you.

Reply:
as in, the resulting slice length is known and can be used by collect() to do a single allocation for the resulting vector?  sorry so pedantic...

Reply:
Yes, the size is exposed to the map iterator, and so it would be surprising if its collect() didn't pre-allocate the destination vector.

Reply:
Also, if needed, using map() opens up additional optimization opportunities including functor deforestation and parallelization.